### ISSUE TITLE
Stateless Graphs

### ISSUE THREAD

**Issue by Gerstenberger** [rwth-i6/returnn#1236]
Currently we are working on converting LMs from TensorFlow to Onnx.
As Onnx does not allow persistent state between session runs, in contrast to TF, we need to manage the LM states explicitly.
For that, we take the initial states as additional input and the last hidden states as additional output.

For LSTMs this was successful using something like this for the placeholders as inputs
```python
 extern_data["lstm0_h"] = {
  "dim": lstm_dim,
  "shape": (lstm_dim,),
  "dtype": "float32",
  "available_for_inference": True}
 extern_data["lstm0_c"] = {
  "dim": lstm_dim,
  "shape": (lstm_dim,),
  "dtype": "float32",
  "available_for_inference": True}
```
and then in the lstm layer: `initial_state": {"h": "data:lstm0_h", "c": "data:lstm0_c"},`

However, for Transformer LMs such an approach does not work.
We get the error:
`invalid initial state type {'k_left': 'data:sa_0_k', 'v_left': 'data:sa_0_v'} for sub-layer 'dec_0_self_att_att', key None` as the `SelfAttentionLayer` calls `RnnCellLayer.get_rec_initial_state_inner` internally to get the initial state.

How to do this for Transformer LMs?

I would suggest to do something similar like `keep_over_epoch` but instead of returning a variable, return a placeholder. Any thoughts on this?

**Comment by albertz** [comment]
> We get the error:
> `invalid initial state type {'k_left': 'data:sa_0_k', 'v_left': 'data:sa_0_v'} for sub-layer 'dec_0_self_att_att', key None` as the `SelfAttentionLayer` calls `RnnCellLayer.get_rec_initial_state_inner` internally to get the initial state.

This sounds like some bug which we should fix anyway, right? Or maybe not bug but rather not yet implemented. But in principle, this should work. Feel free to make a PR with test case and fix for this.

**Comment by albertz** [comment]
> I would suggest to do something similar like `keep_over_epoch` but instead of returning a variable, return a placeholder. Any thoughts on this?

This is another option. This is also ok. Feel free to make a PR for this.

**Comment by albertz** [comment]
Note that with our generalized self-attention implementation (see #391, and `nn.SelfAttention` in returnn-common), all the state handling is explicit, and this would also not really be a problem.

**Comment by albertz** [comment]
How do you do the ONNX conversion? I assume you first create a TF graph, and then convert it somehow? How do you create the TF graph? Do you use `compile_tf_graph.py`? `compile_tf_graph.py` actually also has very similar logic to do what you want, when you use the `--rec_step_by_step` option. Although this uses `tf.Variable`, you could also add an option there to create a `tf.placeholder` instead.

**Comment by Gerstenberger** [comment]
> How do you do the ONNX conversion? I assume you first create a TF graph, and then convert it somehow? How do you create the TF graph? Do you use `compile_tf_graph.py`? `compile_tf_graph.py` actually also has very similar logic to do what you want, when you use the `--rec_step_by_step` option. Although this uses `tf.Variable`, you could also add an option there to create a `tf.placeholder` instead. Feel free to make a PR for this.

yes, i use `compile_tf_graph.py` for this and then call the tf2onnx tool on the resulting graph.

**Comment by albertz** [comment]
> yes, i use `compile_tf_graph.py` for this and then call the tf2onnx tool on the resulting graph.

What options? `--rec_step_by_step`?

**Comment by albertz** [comment]
I guess this can be closed.

You did not answer the last question?

**Comment by Gerstenberger** [comment]
Ah sorry, i forgot.
I don't use `rec_step_by_step` option.

There is some minor things left:
* What would be nice is to have good names for the final states somehow.
Currently i use vanillalstm for the conversion, which produces something like `lstm0/rec/rec_1/while/Exit_4:0` as node names for the h state. For transformer the final states look something like this `output/rec/dec_0_self_att_att/concat:0` for the key state.
How to do this? Insert some tf.identity in the correct places?

* The trafo initial state looks like `output/rec/dec_0_self_att_att/initial_state_placeholder_var:0` for k_left which i can fix later.

**Comment by albertz** [comment]
> I don't use `rec_step_by_step` option.

Why not? This would automatically solve all your problems. This is exactly intended for this use case. We already use it for very similar purpose.



> What would be nice is to have good names for the final states somehow.
> How to do this? Insert some tf.identity in the correct places?

Yea, I guess you can do some similar logic as the `update_var` function for `"keep_over_epoch"`, but not using `assign` but just some `identity` with a name analoge to the placeholder.
